Add "Fence" for delaying initial endpoint regeneration

joamaki opened the pull request:
Traditionally we've used the "CachesSynced" as the mechanism to delay the initial endpoint regeneration. This is now starting to be a difficult mechanism to conform to as we may have a) multiple components consuming the same K8s resource and there can be only one "BlockWaitGroupToSyncResources" per resource, b) we have other data sources besides k8s like kvstore, c) the mechanism is not easy to extend modularly.

This PR proposes adding an "Fence", a wait group style object to which "wait functions" can be added and which can be waited on. This is similar to `StoppableWaitGroup`, but enforces that `Add` is used before Hive starts and `Wait` is used after starting to make sure we don't `Add` to late or `Wait` too early. It also allows bubbling up the error from the initializer through `Wait`.

We provide a "regeneration.Fence" which is just a type wrapper around `hive.Fence` and switch the endpoint "regenerator" to using it. 

The load-balancing and ciliumenvoyconfig waiting is added to this to allow removing of the `BlockWaitGroupToSyncResources` in https://github.com/cilium/cilium/pull/39915.

More context for this is in https://github.com/cilium/cilium/pull/39915#discussion_r2139879852.

joamaki committed:
hive: Add 'Fence' for waiting for initialization

Fence is a simple WaitGroup-style construct for waiting for initialization
of sub-systems. The wait functions can be added only during Hive's population
phase (e.g. when constructors and invoke functions are called) to ensure
that when Wait() is called all possible wait functions have been added.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
loadbalancer: Provide Waitfunc for waiting for LB initialization

Provide a function to wait for LB initialization to finish.

To also wait for redirect policies to be computed, register a LB
table initializer from redirect policies.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
endpoint: Wait for LB initialization before regeneration

Use the [lbcell.InitWaitFunc] to wait for the load-balancing
control-plane to finish initializing before initial regeneration.

We're on purpose doing the Fence.Add from the regenerator
to keep [lbcell.Cell] dependencies small as it's used in integration
tests in many places.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Use regeneration.Fence to block regeneration

Let's be more explicit and block the regeneration with the Fence
instead of going via the "CachesSynced".

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented:
/test

joamaki committed:
endpoint: Use hive.Fence for waiting initializers before regen

Use the [hive.Fence] to make [regeneration.Fence] for dynamically
registering initializers that can block the endpoint regeneration. This provides
an extendable mechanism to add these rather than having to hard-code each one.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/lock/init_waitgroup.go:
```diff
@@ -0,0 +1,93 @@
+	initializers map[string]InitWaitFunc
+}
+
+// Add a named initializer that will block the initial endpoint regeneration until
+// the returned function has been called.
```
Ah the usual blindness to comments. Should perhaps change their color to be brighter so I'm not ignoring them when refactoring.

joamaki commented on pkg/endpoint/regenerator.go:
```diff
@@ -49,30 +53,55 @@ func newRegenerator(in struct {
+		initWaitGroup: in.IWG,
+	}
+
+	// Wait for ipcache sync before regeneration for endpoints including
+	// the ones with fixed identity (e.g. host endpoint), this ensures that
+	// the regenerated datapath always lookups from a ready ipcache map.
+	// Additionally wait for node synchronization, as nodes also contribute
+	// entries to the ipcache map, most notably about the remote node IPs.
+	if option.Config.KVStore != "" {
+		in.IWG.Add(
+			"kvstore",
+			r.waitForKVStoreSync,
+		)
 	}
+
+	// Wait for ipcache and identities synchronization from all remote clusters,
+	// to prevent disrupting cross-cluster connections on endpoint regeneration.
+	in.IWG.Add(
+		"clustermesh",
+		r.waitForClusterMeshIPIdentitiesSync,
+	)
```
Yep. Could we do that in follow-up PR to keep changes here smaller?

squeed commented on pkg/lock/init_waitgroup.go:
```diff
@@ -0,0 +1,94 @@
+//
+// This method will panic if called after Hive has started to ensure that when
+// Wait() is called all initializers have already been registered.
+func (w *InitWaitGroup) Add(name string, waitFn InitWaitFunc) {
```
What about having `Add(name)` return a `doneFunc` that is to be called when setup is complete? Or even just `Add(name)`, `Done(name)`? Might be a bit simpler.

joamaki replied:
This was what I initially did, but then realized most places fit the ”func(Context) error” pattern better. If we’d return a done func we’d need to fork goroutines to wait and call it. We also get nicer stack trace this way.

squeed commented on pkg/endpoint/regeneration/wait.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+// InitWaitGroup for delaying the initial endpoint regeneration.
+type InitWaitGroup interface {
```
Suggestion: expose `lock.InitWaitGroup` interface, then individual timing points just wrap the interface.

squeed commented on pkg/endpoint/regeneration/wait.go:
```diff
@@ -0,0 +1,25 @@
+	Wait(context.Context) error
+}
+
+func NewInitWaitGroup(lc cell.Lifecycle) InitWaitGroup {
```
The naming here is definitely confusing. Maybe we come up with some kind of naming convention like BeforeRegeneration for each synchronization point?

joamaki replied:
Agreed that the naming isn’t great. Will try to come up with something better today.

squeed commented:
Really appreciate bringing order to this particular corner of chaos. I can think of at least 3 timing points we'd want here:

- external state is loaded (kvstore / k8s)
- global bpf maps are written (ipcache, lbmap, etc)
- endpoints have been regenerated

Might be nice to define all these in a single cell for discoverability.

squeed commented:
> Not sure about having a separate "external state is loaded" from "ready to regenerate". Where would we use that coarse signal versus the "per store" signals we have now (e.g. Table[Foo].Initialized, resource.Sync, etc.)?

Hmm, good point. The IPCache waits for k8s caches to be synced before starting its worker thread. Right now it uses `synced.CacheStatus` but I could see this code replacing that mechanism as well, to centralize all timing fences. If we were to get rid of `synced.CacheStatus`, what would you suggest as a replacement? Is there something existing now it should be using?

There are any number of second-order controllers that need to have at least one complete iteration between resources loading and starting endpoint regeneration. Eventually it would be nice to unify the mechanisms.

squeed commented on pkg/hive/fence.go:
```diff
@@ -0,0 +1,109 @@
+// If the context given to it is cancelled the function should stop and return ctx.Err().
+type WaitFunc = func(context.Context) error
+
+// InitWaitGroup is a wait group construct waiting for initialization of a dynamic
```
nit: stale InitWaitGroup reference :-)

joamaki replied:
fixed

joamaki commented:
> > Not sure about having a separate "external state is loaded" from "ready to regenerate". Where would we use that coarse signal versus the "per store" signals we have now (e.g. Table[Foo].Initialized, resource.Sync, etc.)?
> 
> Hmm, good point. The IPCache waits for k8s caches to be synced before starting its worker thread. Right now it uses `synced.CacheStatus` but I could see this code replacing that mechanism as well, to centralize all timing fences. If we were to get rid of `synced.CacheStatus`, what would you suggest as a replacement? Is there something existing now it should be using?
> 
> There are any number of second-order controllers that need to have at least one complete iteration between resources loading and starting endpoint regeneration. Eventually it would be nice to unify the mechanisms.

I would suggest we'd be very explicit about what we wait on. With StateDB we can propagate chains of data dependencies using the table initializers, e.g. if component depends on `Table[A]` and `Table[B]` and writes to `Table[C]` then it'd only mark `Table[C]` as initialized once it has processed A &  B and they've initialized. This way it's much clearer and harder to accidentally miss waiting on something. It also nicely decouples downstream consumers from indirect upstream dependencies.

For existing things not using StateDB I'm not sure we can do much else than keep using `CacheStatus` or some other ad-hoc mechanism.

squeed commented on pkg/endpoint/regenerator.go:
```diff
@@ -45,34 +55,66 @@ func newRegenerator(in struct {
+		fence:         in.Fence,
 	}
+
+	// Wait for ipcache sync before regeneration for endpoints including
```
I understand that this PR is intended to be somewhat minimal, so this makes sense. I would eventually like to see this moved to the individual kvstore / clustermesh cells. Perhaps a `//TODO` to that effect, or a comment not to add any more waits here?

joamaki replied:
Fair enough. Added a comment.

squeed replied:
monkey see, monkey do :-)

squeed reviewed (approved):
Really nice cleanup.

joamaki committed:
daemon: Regenerate restored endpoints after k8s service sync

This is a temporary fix (this "sync-lb-maps-with-k8s-services" will be removed
in #39915) to unblock host endpoint regeneration when IPsec is enabled.

We need to start the "SyncWithK8sFinished" before restored endpoint regeneration
so that the waiting on ClusterMesh to finish doesn't block the regeneration.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented:
/test

jrajahalme reviewed (approved):
This is pretty nice and clean, thanks!

joamaki committed:
loadbalancer: Table initializer in adapter only with k8s

The test/runtime is getting blocked if the adapters register a table initializer
as the "SyncWithK8sFinished" call will only happen when k8s is enabled. So
don't register it if we know k8s is disabled.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

giorio94 commented on pkg/loadbalancer/cell/wait.go:
```diff
@@ -0,0 +1,65 @@
+				return nil
+			}
+
+			// Wait untli frontends table updates. We don't need to rate limit here
```
Nit.

```suggestion
			// Wait until frontends table updates. We don't need to rate limit here
```

giorio94 commented on daemon/cmd/state.go:
```diff
@@ -530,6 +525,11 @@ func (d *Daemon) initRestore(restoredEndpoints *endpointRestoreState, endpointsR
 				syncServices(false /* all services */)
 			}
 		}()
+
+		// When we regenerate restored endpoints, it is guaranteed that we have
+		// received the full list of policies present at the time the daemon
+		// is bootstrapped.
+		d.regenerateRestoredEndpoints(restoredEndpoints, endpointsRegenerator)
```
Nit. This change looks a no-op, and could be reverted.

joamaki replied:
Not a no-op. The ordering matters.

gandro replied:
FWIW, I originally also thought this was a no-op. However, I think `regenerateRestoredEndpoints` now waits for the spawned go-routine give a "sync" signal. But it's not really clear to me either what signal it is.

giorio94 replied:
Chatted offline, I missed that `regenerateRestoredEndpoints` could block if ipsec is enabled, and so we want to ensure that the goroutine is always started before to prevent a deadlock.

giorio94 reviewed (changes requested):
Thanks, looks amazing. Requesting changes based on the comment above. Besides that, just a couple of extra nits inline.

joamaki commented on pkg/hive/fence.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (w *fence) Wait(ctx context.Context) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
```
`x/sync/semaphore` with `Acquire` that takes context OK?
